File: FoodDiary_Frontend/fooddiary-web/src/services/analyticsService.ts
import { API_BASE_URL } from '../constants';
import { getAuthToken } from '../utils';
import type { 
  DailyAnalysis, 
  GetTrendsResponse
} from '../types';

export const analyticsService = {
  
  getDailyAnalysis: async (date: string): Promise<DailyAnalysis> => {
    const token = getAuthToken();
    if (!token) {
      throw new Error('No authentication token available');
    }
    
    const requestBody = { date };
    const response = await fetch(`${API_BASE_URL}/api/analytics/daily`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to fetch daily analysis: ${response.statusText}`);
    }
    
    const data = await response.json();
    
    
    const analysisData = data.analysis || data;
    
    
    const processedData = {
      ...analysisData,
      
      overallStatus: analysisData.overallStatus || 'Poor',
      
      
      isCalorieGoalMet: analysisData.isCalorieGoalMet || false,
      isProteinGoalMet: analysisData.isProteinGoalMet || false,
      isFatGoalMet: analysisData.isFatGoalMet || false,
      isCarbohydrateGoalMet: analysisData.isCarbohydrateGoalMet || false,
      
      
      calorieProgressPercentage: analysisData.calorieProgressPercentage || 0,
      proteinProgressPercentage: analysisData.proteinProgressPercentage || 0,
      fatProgressPercentage: analysisData.fatProgressPercentage || 0,
      carbohydrateProgressPercentage: analysisData.carbohydrateProgressPercentage || 0,
      
      
      totalCalories: analysisData.totalCalories || 0,
      totalProtein: analysisData.totalProtein || 0,
      totalFat: analysisData.totalFat || 0,
      totalCarbohydrates: analysisData.totalCarbohydrates || 0,
      totalFoodEntries: analysisData.totalFoodEntries || 0,
      
      
      dailyCalorieGoal: analysisData.dailyCalorieGoal || 0,
      dailyProteinGoal: analysisData.dailyProteinGoal || 0,
      dailyFatGoal: analysisData.dailyFatGoal || 0,
      dailyCarbohydrateGoal: analysisData.dailyCarbohydrateGoal || 0,
      
      
      isOverCalorieLimit: analysisData.isOverCalorieLimit || false,
      isOverProteinLimit: analysisData.isOverProteinLimit || false,
      isOverFatLimit: analysisData.isOverFatLimit || false,
      isOverCarbohydrateLimit: analysisData.isOverCarbohydrateLimit || false,
      
      
      recommendations: analysisData.recommendations?.map((rec: string, index: number) => {
        const text = rec.toLowerCase();
        let type = 'info';

        
        if (text.includes('varies significantly') || text.includes('varies day to day')) {
          type = 'consistency';
        }
        
        else if (text.includes('meals per day') || text.includes('meal timing') || text.includes('meal prep')) {
          type = 'meal-pattern';
        }
        
        else if (text.includes('rarely meet') || text.includes('sometimes miss')) {
          type = 'goal';
        }
        
        else if (text.includes('highly inconsistent') || text.includes('moderate inconsistency')) {
          type = 'consistency';
        }
        
        else if (text.includes('weight loss') || text.includes('weight gain') || text.includes('maintenance')) {
          type = 'fitness';
        }

        return {
          index,
          text: rec,
          type
        };
      }) || []
    };
    

    
    return processedData;
  },

  
  getTrends: async (days: number = 7): Promise<GetTrendsResponse> => {
    const token = getAuthToken();
    if (!token) {
      throw new Error('No authentication token available');
    }
    
    const requestBody = { days };
    
    const response = await fetch(`${API_BASE_URL}/api/analytics/trends`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to fetch trends: ${response.statusText}`);
    }
    
    const data = await response.json();
    
    
    const { trends, recommendations, message } = data;
    
    
    const processedRecommendations = recommendations?.map((rec: string, index: number) => {
      const text = rec.toLowerCase();
      let type = 'info';

      
      if (text.includes('varies significantly') || text.includes('varies day to day') || 
          text.includes('highly inconsistent') || text.includes('inconsistent')) {
        type = 'consistency';
      }
      
      else if (text.includes('meals per day') || text.includes('meal timing') || text.includes('meal prep')) {
        type = 'meal-pattern';
      }
      
      else if (text.includes('rarely meet') || text.includes('sometimes miss') || text.includes('meet your')) {
        type = 'goal';
      }
      
      else if (text.includes('weight loss') || text.includes('weight gain') || text.includes('maintenance')) {
        type = 'fitness';
      }

      return {
        index,
        text: rec,
        type
      };
    });
    
    
    const processedTrends = {
      ...trends,
      
      averageCalories: trends.averageCalories || 0,
      averageProtein: trends.averageProtein || 0,
      averageFat: trends.averageFat || 0,
      averageCarbohydrates: trends.averageCarbohydrates || 0,
      
      calorieConsistency: trends.calorieConsistency || 0,
      proteinConsistency: trends.proteinConsistency || 0,
      fatConsistency: trends.fatConsistency || 0,
      carbohydrateConsistency: trends.carbohydrateConsistency || 0,
      
      goalAdherenceRate: (trends.goalAdherenceRate || 0) / 100, // Convert percentage to decimal (0-1)
      daysGoalsMet: trends.daysGoalsMet || 0,
      
      averageMealsPerDay: trends.averageMealsPerDay || 0,
      mostCommonMealTime: trends.mostCommonMealTime || 'Not enough data',
      leastCommonMealTime: trends.leastCommonMealTime || 'Not enough data',
      
      calorieTrend: trends.calorieTrend || 'Stable',
      proteinTrend: trends.proteinTrend || 'Stable',
      fatTrend: trends.fatTrend || 'Stable',
      carbohydrateTrend: trends.carbohydrateTrend || 'Stable',
      overallTrend: trends.overallTrend || 'Stable',
      
      isConsistent: trends.isConsistent || false,
      isImproving: trends.isImproving || false,
      trendInsights: trends.trendInsights || []
    };
    

    
    return { 
      trends: processedTrends, 
      recommendations: processedRecommendations, 
      message 
    };
  },
};

export default analyticsService;

File: FoodDiary_API/src/FoodDiary.UseCases/Analytics/AnalyticsService.cs
using FoodDiary.Core.AnalyticsAggregate;

namespace FoodDiary.UseCases.Analytics;

public interface IAnalyticsService
{
    bool IsCalorieGoalMet(DailyNutritionalAnalysis analysis);
    bool IsProteinGoalMet(DailyNutritionalAnalysis analysis);
    bool IsFatGoalMet(DailyNutritionalAnalysis analysis);
    bool IsCarbohydrateGoalMet(DailyNutritionalAnalysis analysis);
    
    bool IsOverCalorieLimit(DailyNutritionalAnalysis analysis);
    bool IsOverProteinLimit(DailyNutritionalAnalysis analysis);
    bool IsOverFatLimit(DailyNutritionalAnalysis analysis);
    bool IsOverCarbohydrateLimit(DailyNutritionalAnalysis analysis);
    
    string GetOverallStatus(DailyNutritionalAnalysis analysis);
    Task<string> GetOverallStatusAsync(DailyNutritionalAnalysis analysis);
}

public class AnalyticsService : IAnalyticsService
{
    private const double GoalAchievementThreshold = 0.8;
    private const double ExcellentThreshold = 90.0;
    private const double GoodThreshold = 75.0;
    private const double FairThreshold = 60.0;
    private const double NeedsImprovementThreshold = 40.0;

    public bool IsCalorieGoalMet(DailyNutritionalAnalysis analysis) =>
        IsGoalMet(analysis.DailyCalorieGoal, analysis.TotalCalories);

    public bool IsProteinGoalMet(DailyNutritionalAnalysis analysis) =>
        IsGoalMet(analysis.DailyProteinGoal, analysis.TotalProtein);

    public bool IsFatGoalMet(DailyNutritionalAnalysis analysis) =>
        IsGoalMet(analysis.DailyFatGoal, analysis.TotalFat);

    public bool IsCarbohydrateGoalMet(DailyNutritionalAnalysis analysis) =>
        IsGoalMet(analysis.DailyCarbohydrateGoal, analysis.TotalCarbohydrates);

    public bool IsOverCalorieLimit(DailyNutritionalAnalysis analysis) =>
        IsOverLimit(analysis.DailyCalorieGoal, analysis.TotalCalories);

    public bool IsOverProteinLimit(DailyNutritionalAnalysis analysis) =>
        IsOverLimit(analysis.DailyProteinGoal, analysis.TotalProtein);

    public bool IsOverFatLimit(DailyNutritionalAnalysis analysis) =>
        IsOverLimit(analysis.DailyFatGoal, analysis.TotalFat);

    public bool IsOverCarbohydrateLimit(DailyNutritionalAnalysis analysis) =>
        IsOverLimit(analysis.DailyCarbohydrateGoal, analysis.TotalCarbohydrates);

    public string GetOverallStatus(DailyNutritionalAnalysis analysis)
    {
        var (metGoals, totalGoals) = CalculateGoalAchievement(analysis);
        
        if (totalGoals == 0) return "No goals set";
        
        var percentage = CalculateAchievementPercentage(metGoals, totalGoals);
        return DetermineStatusFromPercentage(percentage);
    }

    public async Task<string> GetOverallStatusAsync(DailyNutritionalAnalysis analysis) =>
        await Task.Run(() => GetOverallStatus(analysis));

    private static bool IsGoalMet(double? goal, double actual) =>
        goal.HasValue && actual >= goal.Value * GoalAchievementThreshold;

    private static bool IsOverLimit(double? goal, double actual) =>
        goal.HasValue && actual > goal.Value * 1.1;

    private static (int metGoals, int totalGoals) CalculateGoalAchievement(DailyNutritionalAnalysis analysis)
    {
        var metGoals = 0;
        var totalGoals = 0;
        
        if (analysis.DailyCalorieGoal.HasValue)
        {
            totalGoals++;
            if (IsGoalMet(analysis.DailyCalorieGoal, analysis.TotalCalories) && 
                !IsOverLimit(analysis.DailyCalorieGoal, analysis.TotalCalories)) 
                metGoals++;
        }
        
        if (analysis.DailyProteinGoal.HasValue)
        {
            totalGoals++;
            if (IsGoalMet(analysis.DailyProteinGoal, analysis.TotalProtein)) 
                metGoals++;
        }
        
        if (analysis.DailyFatGoal.HasValue)
        {
            totalGoals++;
            if (IsGoalMet(analysis.DailyFatGoal, analysis.TotalFat)) 
                metGoals++;
        }
        
        if (analysis.DailyCarbohydrateGoal.HasValue)
        {
            totalGoals++;
            if (IsGoalMet(analysis.DailyCarbohydrateGoal, analysis.TotalCarbohydrates)) 
                metGoals++;
        }
        
        return (metGoals, totalGoals);
    }

    private static double CalculateAchievementPercentage(int metGoals, int totalGoals) =>
        (double)metGoals / totalGoals * 100;

    private static string DetermineStatusFromPercentage(double percentage) =>
        percentage switch
        {
            >= ExcellentThreshold => "Excellent",
            >= GoodThreshold => "Good",
            >= FairThreshold => "Fair",
            >= NeedsImprovementThreshold => "Needs Improvement",
            _ => "Poor"
        };
} 